drafts/draft-svaldez-pp-http-api/draft-svaldez-pp-http-api.md

```diff
@@ -0,0 +1,423 @@
+    opaque verification_key<1..2^16-1>;
+    KeyCommitment commitments<1..2^16-1>;
+}
+~~~
+
+5. The client then verifies the signature for each key commitment and
+stores the list of commitments to the current scope. The client SHOULD
+not cache the commitments beyond the current scope, as new commitments
+should be fetched for each independent issuance and redemption request.
+
+# Privacy Pass Issuance {#issuance}
+
+Inputs:
+- ``server_origin``: The origin to request token issuance from.
+- ``count``: The number of tokens to request issuance for.
+
+Outputs:
+- ``tokens``: A list of tokens that have been signed via the Privacy
+  Pass protocol.
+
+1. When a client wants to request tokens from a server, it should first
+   fetch a key commitment from the issuer via the process described in
+   {{key-commitment}} and keep the result as ``commitment``.
+
+2. The client should then call the ``CLIENT_ISSUE_GEN`` interface
+   requesting ``count`` tokens storing the resulting ``c_data``,
+   ``i_data``, ``g_data``.
+
+3. The client then makes a POST request to
+   <``server_origin``>/.well-known/privacy-pass with a message of type
+   ``request-issuance`` and a body of:
+
+~~~
+enum { Normal(0) } IssuanceType;
+
+struct {
+    IssuanceType type = 0;
+    opaque issue_data<0..2^16-1> = client_issue.issue_data;
+}
+~~~
+
+4. The server, upon receipt of the ``request`` should call the
+   ``SERVER_ISSUE`` interface with the value of ``issue_data`` with a
+   result of ``server_issue_resp``.
+
+
+5. The server should then respond to the POST request with a message of
+   type ``issue`` and a body of:
+
+~~~
+struct {
+    IssuanceType type = request.type;
+    uint16 ciphersuite = server_issue_resp.ciphersuite;
+    opaque evals<1..2^16-1> = server_issue_resp.evals;
+    opaque proof<1..2^16-1> = server_issue_resp.proof;
+}
+~~~
+
+6. The client should then should call the ``CLIENT_ISSUE_FINISH``
+   interface with the resulting ``ciphersuite``, ```evals``, ``proof``,
+   and stored ``g_data`` to extract a list of ``tokens``.
+
+7. The client should store the ``public_key`` associated with these
+   tokens and the elements of ``tokens`` under storage partitioned by
+   the ``server_origin``, accessible only via the Privacy Pass API.
+
+# Privacy Pass Redemption {#redemption}
+
+There are two forms of Privacy Pass redemption that could function under
+the HTTP API. Either passing along a token directly to the target
+endpoint, which would perform its own redemption
+{{token-redemption}}, or the client redeeming the token and
+passing the result along to the target endpoint. These two methods are
+described below.
+
+## Generic Token Redemption {#token-redemption}
+
+Inputs:
+- ``server_id``: The server ID to redeem a token against.
+- ``ciphersuite``: The ciphersuite for this token.
+- ``public_key``: The public key associated with this token.
+- ``token``: A Privacy Pass token.
+- ``additional_data``: Additional data to bind to this token redemption.
```

Comment:
Eventually we may want to be more prescriptive about how additional_data is formed as it is the HMAC input. This document may be the best place to do that.

Reply:
Specific construction for the delegated case is down below, but I'm not sure whether the client (UA) which originally has the tokens should be doing PP_Redeem and then passing the result of that to whatever website wants to redeem a token (allowing the UA to set the aux additional_data, so that it can't be reused elsewhere), or whether the website should be taking the raw token and doing its own PP_Redeem with whatever it wants to insert in aux.